bam_addrprg.c

```diff
@@ -0,0 +1,461 @@
+            free(tmp);
+            return false;
+        }
+        retval->rg_line = tmp;
+    }
+    retval->input_name = strdup(argv[optind+0]);
+    retval->output_name = strdup(argv[optind+1]);
+    retval->mode = mode;
+
+    if (sam_open_mode(&modeout[1], retval->output_name, fmtout) < 0) {
+        if (fmtout) fprintf(stderr, "[bam_addrprg] can't parse output format \"%s\"\n", fmtout);
+        else fprintf(stderr, "[bam_addrprg] can't determine output format\n");
+        return false;
+    }
+    retval->output_mode = strdup(modeout);
+
+    if (level >= 0) sprintf(strchr(modeout, '\0'), "%d", level < 9? level : 9);
+
+    return true;
+}
+
+static void overwrite_all_func(const state_t* state, bam1_t* file_read)
+{
+    uint8_t* data = (uint8_t*)strdup(state->rg_id);
+    int len = strlen(state->rg_id)+1;
+    // If the old exists delete it
+    uint8_t* old = bam_aux_get(file_read, "RG");
+    if (old != NULL) {
+        printf("deleting\n");
+        bam_aux_del(file_read, old);
+    }
+
+    bam_aux_append(file_read, "RG",'Z',len,data);
+    free(data);
+}
+
+static void orphan_only_func(const state_t* state, bam1_t* file_read)
+{
+    uint8_t* data = (uint8_t*)strdup(state->rg_id);
+    int len = strlen(state->rg_id)+1;
+    // If the old exists don't do anything
+    uint8_t* old = bam_aux_get(file_read, "RG");
+    if (old == NULL) {
+        bam_aux_append(file_read, "RG",'Z',len,data);
+    }
+    free(data);
+}
+
+static bool init(parsed_opts_t* opts, state_t** state_out) {
+    state_t* retval = (state_t*) calloc(1, sizeof(state_t));
+    if (retval == NULL) {
+        fprintf(stderr, "Out of memory\n");
+        return false;
+    }
+    *state_out = retval;
+
+    // Open files
+    retval->input_file = sam_open(opts->input_name, "r");
+    if (retval->input_file == NULL) {
+        fprintf(stderr, "Could not open input file: %s\n", opts->input_name);
+        return false;
+    }
+    retval->input_header = sam_hdr_read(retval->input_file);
+
+    retval->output_header = bam_hdr_dup(retval->input_header);
+    retval->output_file = sam_open(opts->output_name, opts->output_mode);
+    
+    if (retval->output_file == NULL) {
+        print_error_errno("Could not open output file: %s mode: %s\n", opts->output_name, opts->output_mode);
+        return false;
+    }
+    
+    if (opts->rg_line) {
+        // Append new RG line to header.
+        // Check does not already exist
+        if ( confirm_rg(retval->output_header, opts->rg_id) ) {
+            fprintf(stderr, "ID of new RG line specified conflicts with that of an existing header RG line. Overwrite not yet implemented.\n");
+            return false;
+        }
+        retval->rg_id = strdup(opts->rg_id);
+        size_t new_len = strlen( retval->output_header->text ) + strlen( opts->rg_line ) + 2;
+        char* new_header = malloc(new_len);
+        if (!new_header) {
+            fprintf(stderr, "malloc error when writing new header.\n");
```

Comment:
Nitpicking now - but your other memory related errors go with a simpler and more user-friendly "Out of memory [whilst allocating <x>]".